### ISSUE TITLE
Rest API with FileUpload (Form-data)

### ISSUE THREAD

**Issue by abhishek-27** [kelaberetiv/TagUI#454]
Hello Kensoh,

I am facing difficulty in using the API step for a file upload API. 
It's a **POST** API and I have set the **api_config** variable with the details. 
```
api_config = {method:'POST', header:['apikey': '**********'], body:{'fileUploaded':'/Users/test/Downloads/Test.pdf'}};

api https://http://www.****.com/****/aa5474b9-fb06-4d87-8604-dc731db15290

echo api_json

```

So from Postman App, it's pretty straightforward but I am not able to understand how to pass the file path in your API declaration.

**Comment by kensoh** [comment]
Hi @abhishek-27 your code looks correct, what error messages do you get after making the API call?

In the backend, TagUI will use below to convert the api_config.body you provided into JSON string, before sending it through a XMLHttpRequest.

```js
JSON.stringify(api_config.body)
```

Alternatively, as TagUI has integration with Python, consider using the Python requests package to do the API call and print the result back to be used as py_result variable. Details here - https://github.com/kelaberetiv/TagUI/issues/448#issuecomment-500529402

**Comment by abhishek-27** [comment]
Thanks, Ken.

The error I got was related to Phantom JS. 

```
SyntaxError: Expected token ']'

  phantomjs://code/new.js:1399 in injectJs
  phantomjs://code/bootstrap.js:456
```

Yes so as an alternative I used Python but then while importing requests I was again getting an error, seemed like it wasn't allowing to import third-party libraries. So, for now, I used subprocess and curl command to make the API call.

One more issue which I was facing was in the UI, I had a dropdown and based on the documentation I tried a few things but then the option value is not getting selected.

`select //*[@id="abc1"]/div/div/div/div[1]/div/div/div/div[3]/div[1]/div/span[1]/div/select/ as Travelling
`
And even tried specifying the option number instead.
`select //*[@id="abc1"]/div/div/div/div[1]/div/div/div/div[3]/div[1]/div/span[1]/div/select/ as 2
` or `select //*[@id="abc1"]/div/div/div/div[1]/div/div/div/div[3]/div[1]/div/span[1]/div/select/ as option[2]`

The documentation didn't have many examples so was a little hard to follow. Can you help me out here?

Thanks.

**Comment by abhishek-27** [comment]
One more thing I wanted to know was how can I use popup step to actually open a new webpage in . anew tab with a totally new URL. 

I tried `popup new_url { http://www.google.com }`.
This didn't work neither this `popup  http://www.google.com`.

**Comment by kensoh** [comment]
Oh great, thanks for the error message. Looks like there is a small typo in your API call which I also missed. From the readme doc, the header is defined as follows. The header value pair is actually a single string.
```js
api_config = {method:'PUT', header:['Header1: value1','Header2: value2'], body:{'id':123,'pwd':'abc'}};
```

So your api call should be as follows instead

```js
api_config = {method:'POST', header:['apikey: **********'], body:{'fileUploaded':'/Users/test/Downloads/Test.pdf'}};
```

If API key is a variable, you can send it as follows

```js
api_config = {method:'POST', header:['apikey:'+ apikey_variable], body:{'fileUploaded':'/Users/test/Downloads/Test.pdf'}};
```

You can try again to see if above suggestion works for you!

For the Python import, it should work if in your terminal the normal Python interactive shell you can do import. After doing the pip install requests and running `python` command and doing `import requests`. It may be some Python environment setup issue.

For your 2 other questions, I'll raise 2 new issues so that it is easier for other users to learn from the answers when they search.

**Comment by kensoh** [comment]
Issue for your question of selecting option value - https://github.com/kelaberetiv/TagUI/issues/455

**Comment by kensoh** [comment]
Issue for your question of using popup step - https://github.com/kelaberetiv/TagUI/issues/456
